network/gopubsub/pubsub.go

```diff
@@ -0,0 +1,126 @@
+package gopubsub
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+)
+
+//PubSubChannel represents a pubsub system where subscriber can .Subscribe() and publishers can .Publish() or .Close().
+type PubSubChannel struct {
+	subscriptions []*Subscription
+	subsMutex     sync.RWMutex
+	nextId        int
+}
+
+//Subscription is created in PubSubChannel using pubSub.Subscribe(). Subscribers can receive using .ReceiveChannel().
+// or .Unsubscribe()
+type Subscription struct {
+	closed    bool
+	removeSub func()
+	signal    chan interface{}
+	closeOnce sync.Once
+	id        string
+}
+
+//New creates a new PubSubChannel.
+func New() *PubSubChannel {
+	return &PubSubChannel{
+		subscriptions: make([]*Subscription, 0),
+	}
+}
+
+//Subscribe creates a subscription to a channel, each subscriber should keep its own Subscription instance.
+func (psc *PubSubChannel) Subscribe() *Subscription {
+	psc.subsMutex.Lock()
+	defer psc.subsMutex.Unlock()
+	newSubscription := newSubscription(strconv.Itoa(psc.nextId))
+	psc.nextId++
+	psc.subscriptions = append(psc.subscriptions, &newSubscription)
+	newSubscription.removeSub = func() {
+		psc.subsMutex.Lock()
+		defer psc.subsMutex.Unlock()
+
+		for i, subscription := range psc.subscriptions {
+			if subscription.signal == newSubscription.signal {
+				fmt.Println("Unsubscribing", "id", subscription.id)
+				subscription.closed = true
+				psc.subscriptions = append(psc.subscriptions[:i], psc.subscriptions[i+1:]...)
+			}
+		}
+	}
+	return &newSubscription
+}
+
+//Publish broadcasts a message asynchronously to each subscriber.
+//If some of the subscriptions(channels) has been marked as closeable, it does it now.
+func (psc *PubSubChannel) Publish(msg interface{}) {
+	psc.subsMutex.RLock()
+	defer psc.subsMutex.RUnlock()
+	for i, sub := range psc.subscriptions {
+		if sub.closed {
+			fmt.Println("Subscription was closed", "id", sub.id)
+			sub.closeChannel()
```

Comment:
Why is channel closed here in Publish and not on Unsubscribe or removeSub? Any subscription may wait for the next publish to receive a false from the channel, which may not happen and the receiver will wait. Also, there would be no need for closed bool field, except for unused IsClosed method.

Reply:
No, the receiver don't wait for the channel to be closed. 
I could add a tick in the PubSubChannel to close periodically the sunscriptions but I think is not necessary. The publisher will close it next time a message is published, and if not, all channel will be closed on PubSubChannel disposal (Close() function).

Reply:
I think that channel closing is not a blocking operation. But, I do not think that adding more complexity with ticker is needed.